File: DSA/Shubham_590016506/Day_10_Question_2.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

int main() {
    char s[200], all[200], trim[200], single[200];
    fgets(s, sizeof(s), stdin);
    int i = 0, j = 0;
    while (s[i]) if (s[i] != ' ') all[j++] = s[i++]; else i++;
    all[j] = '\0';

    i = 0;
    while (isspace(s[i])) i++;
    j = strlen(s) - 1;
    while (j >= 0 && isspace(s[j])) j--;
    strncpy(trim, s + i, j - i + 1);
    trim[j - i + 1] = '\0';

    i = 0, j = 0;
    int space = 0;
    while (s[i]) {
        if (!isspace(s[i])) {
            single[j++] = s[i++];
            space = 0;
        } else {
            if (!space) single[j++] = ' ';
            space = 1;
            i++;
        }
    }
    single[j] = '\0';
    i = 0;
    while (isspace(single[i])) i++;
    j = strlen(single) - 1;
    while (j >= 0 && isspace(single[j])) j--;
    single[j + 1] = '\0';

    printf("Remove all: \"%s\", Trim: \"%s\", Single spaces: \"%s\"\n", all, trim, single);
    return 0;
}

File: DSA/Sauhard_Semalty_590013630/Day10Question2.py
import sys


def remove_all_spaces(string: str) -> str:
    "Removes all spaces from a string"

    result = ""
    for c in string:
        if not c.isspace():
            result += c
    return result


def trim(string: str) -> str:
    "Removes leading and trailing spaces from a string"

    leading = 0
    trailing = 0

    for index, c in enumerate(string):
        if not c.isspace():
            leading = index
            break
    else:
        return ""  # The string does not contain any characters

    for index, c in enumerate(string[::-1]):
        if not c.isspace():
            trailing = len(string) - index
            break


    return string[leading:trailing]


def remove_extra_space(string: str) -> str:
    "Removes extra space around words in a string"

    result = ""
    word = ""

    for c in string + " ":
        if not c.isspace():
            word += c
        else:
            if word:
                result += word + " "
                word = ""

    return result[:-1]  # Remove the extra space at the end


def main() -> None:
    "The entry point of the program."

    string = input("Enter string: ")
    if not string:
        print("Empty string is not allowed.")
        sys.exit(-1)

    print(f"Remove all: {remove_all_spaces(string)!r}, Trim: {trim(string)!r}"
          f", Single spaces: {remove_extra_space(string)!r}")


if __name__ == "__main__":
    main()
